src/korg_nanokontrol/korg_nanokontrol.py

```diff
@@ -0,0 +1,154 @@
+
+        self.msg = Joy()
+        self.mode = None
+
+        self.pub = rospy.Publisher('joy', Joy, latch=True)
+
+    def update(self):
+        ''' run the update hook (poll the midi controller, publish the message)
+        '''
+        while self.controller.poll():
+            data = self.controller.read(1)
+            if self.mode:
+                self.pub_data(data)
+            else:
+                self.set_mode(self.guess_mode(data))
+
+    def set_mode(self, mode):
+        ''' Set up the message for the correct mode
+        '''
+        if mode:
+            self.mode = mode
+            self.msg.axes = [0]*len(self.modes[mode]['control_axis'])
+            self.msg.buttons = [0]*len(self.modes[mode]['control_buttons'])
+            self.msg.buttons.append(mode)
+
+    def guess_mode(self, data):
+        ''' Given data, try to guess the mode we are in
+        '''
+        for event in data:
+            control = event[0]
+
+            # look for continuous controller commands
+            if (control[0] & 0xF0) == 176:
+                control_id = control[1] | ((control[0] & 0x0F) << 8)
+
+                # guess initial mode based on command
+                candidate = None
+                for index, control_mode in enumerate(self.modes):
+                    if control_id in control_mode['control_axis']:
+                        if candidate is not None:
+                            candidate = None
+                            break
+                        candidate = index
+
+                    if control_id in control_mode['control_buttons']:
+                        if candidate is not None:
+                            candidate = None
+                            break
+                        candidate = index
+                if not candidate:
+                    rospy.loginfo("determining mode..")
+                return candidate
+
+    def pub_data(self, data):
+        ''' Given data, interpret into Joy message and publish
+        '''
+        do_publish = False
+        for event in data:
+            control = event[0]
+            # @todo use timestamp from msg
+            # timestamp = event[1]
+
+            # look for continuous controller commands
+            if (control[0] & 0xF0) == 176:
+                control_id = control[1] | ((control[0] & 0x0F) << 8)
+
+                control_axis = self.modes[self.mode]['control_axis']
+                control_buttons = self.modes[self.mode]['control_buttons']
+
+                if control_id in control_axis:
+                    if self.center_axis:
+                        control_val = float(control[2] - 63) / 63.0
+                        if control_val < -1.0:
+                            control_val = -1.0
+                        if control_val > 1.0:
+                            control_val = 1.0
+                    else:
+                        control_val = float(control[2]) / 127
+                        if control_val < 0:
```

Comment:
good catch!

Is numpy too much of an overhead to include?  Then I could just use numpy.clip()